# Question 1

Given two strings s and t, determine whether some anagram of t is a substring
of s. For example: if s = "udacity" and t = "ad", then the function returns
True. Your function definition should look like: question1(s, t) and return a
boolean True or False.


In [5]:
# Turn on/off to enable/disable debugging
debug = False


def is_anagram(s, t):
    """
    # Find if strings are anagram. t anagram of s
    # @param {string, string} input strings
    # @return bool if strings are anagram of each other or not
    """
    s = list(s)
    # Sort a string and then compare with each other
    s.sort()   # Quick sort O(n*log(n))
    return s == t


def question1(s,t):
    """
    # Find out sorted(possible substring of s) and compare with sorted(t)
    # @param main_string, ana_string input strings
    # @return bool Question1 answer
    """
    global debug
    sort_t = list(t)
    sort_t.sort()     # Quick sort O(n*log(n))

    for i in range(len(s) - len(t) + 1):
        if debug:
            print (s[i: i+len(t)], t)
        if is_anagram(s[i: i+len(t)], sort_t):
            return True
    return False

# Main program
def main():
    unit_tests = [("udacity", "ad"),
                 ("udacity", "da"),
                 ("udacity", "au"),
                 ("udacity", "ciy"),
                 ("udacity", "xyz"),
                 ("udacity", ""),
                 ("udacity", "12863742302323232746"),
                 ("ad", "udacity" ),
                 ("panama", "1pana"),
                 ("banana", "anana")]

    for test in unit_tests:
        print("{} -> {}".format(test,question1(test[0],test[1])))


if __name__ == '__main__':
    main()

('udacity', 'ad') -> True
('udacity', 'da') -> True
('udacity', 'au') -> False
('udacity', 'ciy') -> False
('udacity', 'xyz') -> False
('udacity', '') -> True
('udacity', '12863742302323232746') -> False
('ad', 'udacity') -> False
('panama', '1pana') -> False
('banana', 'anana') -> True


## Answer

Given the test cases bellow:

#### Case 1: ("udacity", "ad") -> True
string with 2 characters reversed from order in S

####  Case 2: ("udacity", "da") -> True
string with 2 characters same as in order in S

####  Case 3: ("udacity", "au") -> False
string with 2 alternative characters and reversed from order in S

####  Case 4: ("udacity, dacity") -> True
string with 6 characters same as in order in S

####  Case 5: ("udacity", "ciy") -> False
string with 2 characters same as in order in S and 1 character out of order.

####  Case 6: ("udacity", "xyz") -> False
string with new characters not in string S.

#### Case 7: ("udacity", "") -> True
empty string.

####  Case 8: ("ad", "udacity" ) -> False
string t's length is greater than string S.

####  Case 9: ("panama", "1pana" ) -> False
string t's contains a number before.

####  Case 10: ("banana", "anana" ) -> False
string t's is an anagram of string S

All tests are passed as it can be seen above.


# Question 2
Given a string a, find the longest palindromic substring contained in a.
Your function definition should look like question2(a), and return a string.


In [6]:
def isPalindrome(string):
    """
    # @param string s input string
    # @return bool if string is palindrome or not
    """
    if not string:
        return False
    # reverse compare
    return string == string[::-1]


def question2(string):
    """
    # @param string s input string
    # @return string the longest palindromic substring
    """
    if not string: # if null entry
        return ""

    n = len(string)
    longest, left, right = 0, 0, 0
    for i in range(0, n):
        for j in range(i + 1, n + 1):
            substr = string[i:j] # going through all substrings possible
            if isPalindrome(substr) and len(substr) > longest: # see if it is palindrome and is longer than the one before
                longest = len(substr)
                left, right = i, j
    # construct longest substr
    result = string[left:right]
    return result

# Main program
def main():
    unit_tests = ["abuttuba",
                 "acaramanamaraca",
                 "adogapanicinapagoda",
                 "anutforajaroftuna",
                 "asantaatnasa",
                 "amoreroma",
                 "ABCBACADCBBCDA",
                 "iriwmairanariairie",
                 "anaannaanne",
                 "",
                 "1234432"]

    for test in unit_tests:
        print("{} -> {}".format(test,question2(test)))



if __name__ == '__main__':
    main()


abuttuba -> abuttuba
acaramanamaraca -> acaramanamaraca
adogapanicinapagoda -> adogapanicinapagoda
anutforajaroftuna -> anutforajaroftuna
asantaatnasa -> asantaatnasa
amoreroma -> amoreroma
ABCBACADCBBCDA -> ADCBBCDA
iriwmairanariairie -> airanaria
anaannaanne -> naannaan
 -> 
1234432 -> 234432


## Answer

Given the test cases bellow:

#### Case 1: A But Tuba
#### Case 2: A car, a man, a maraca
#### Case 3: A dog! A panic in a pagoda
#### Case 4: A nut for a jar of tuna
#### Case 5: A Santa at Nasa.
#### Case 6: Amore, Roma.
#### Case 7: ABCBAC ADCBBCDA
#### Case 8: iriwm airanaria irie
#### Case 9: a naannaan ne
#### Case 10: null entry
#### Case 11: 1 234432 

All tests are passed as it can be seen above.
